### PR TITLE
Pytorch parallelism

### LINKED ISSUE #5871: [Concept Entry] PyTorch: Distributed Data Parallelism
### Reason/inspiration (optional)

We would like a new entry on `Distributed Data Parallelism` concept under PyTorch. The entry should go in a new file under `docs/content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md`.       
  
The entry should include:      
  
- An introduction to the concept      
- A `Syntax` section that provides the syntax for the concept (if applicable)      
- An `Example` section that provides an example demonstrating the concept in use (if applicable)     

 
Please refer to the [concept entry template](https://github.com/Codecademy/docs/blob/main/documentation/concept-entry-template.md), [content standards](https://github.com/Codecademy/docs/blob/main/documentation/content-standards.md), and [markdown style guide](https://curriculum-documentation.codecademy.com/content-guidelines/markdown-style-guide/) when working on the PR for this issue.|

### Entry Type (select all that apply)

New Concept Entry for Existing Topic

### Code of Conduct

- [X] By submitting this issue, I agree to follow Codecademy Doc's [Code of Conduct](https://github.com/Codecademy/docs/blob/main/.github/CODE_OF_CONDUCT.md).

### For Maintainers

- [X] Labels added
- [ ] Issue is assigned

### PR DESCRIPTION
**Pull request by andersooi**
### Description

- Added new concept entry for PyTorch Distributed Data Parallelism

### Issue Solved

Closes #5871 

### Type of Change

- Adding a new entry

### Checklist

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [x] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [x] I have linked any issues that are relevant to this PR in the `Issues Solved` section.

### CODE REVIEW

File: content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
+---
+Title: 'Distributed Data Parallelism'
+Description: 'An overview of distributed data parallelism in PyTorch.'
```
Comment (Radhika-okhade) [inline content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md:3]: The description is too generic. Explain what distributed data parallelism is within 1-2 lines. 
https://github.com/Codecademy/docs/blob/main/documentation/style-guide.md

File: content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+Subjects:
+  - 'Data Science'
+  - 'Machine Learning'
+Tags:
+  - 'PyTorch'
+  - 'Data'
+  - 'Data Parallelism'
```
Comment (Radhika-okhade) [inline content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md:10]:
```suggestion
  - 'Data'
  - 'Data Parallelism'
  - 'PyTorch'
```

File: content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+CatalogContent:
+  - 'intro-to-py-torch-and-neural-networks'
+  - 'paths/build-a-machine-learning-model'
+---
+
+## Introduction to Distributed Data Parallelism
+
+Distributed Data Parallelism (DDP) in PyTorch is a module that enables users to efficiently train models across multiple GPUs and machines. By splitting the training process across multiple machines, DDP helps reduce training time and facilitates scaling to larger models and datasets. It achieves parallelism by splitting the input data into smaller chunks, processing them on different GPUs, and aggregating results for updates. Compared to `DataParallel`, DDP offers better performance and scalability by minimising device communication overhead.
```
Comment (Radhika-okhade) [inline content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md:18]:
```suggestion
Distributed Data Parallelism (DDP) in PyTorch is a module that enables users to efficiently train models across multiple GPUs and machines. By splitting the training process across multiple machines, DDP helps reduce training time and facilitates scaling to larger models and datasets. 
It achieves parallelism by splitting the input data into smaller chunks, processing them on different GPUs, and aggregating results for updates. Compared to `DataParallel`, DDP offers better performance and scalability by minimizing device communication overhead.
```

File: content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+- High performance: Efficient communication using NVIDIA Collective Communications Library (NCCL) or Gloo backend (for Windows platforms).
+- Fault tolerance: Handles errors during distributed training.
+- Scalability: Suitable for multi-node, multi-GPU setups.
+
+## Syntax
+
+To use DDP, a distributed process group needs to be initialised and wrapped to a model with `torch.nn.parallel.DistributedDataParallel`.
+
+```py
```
Comment (Radhika-okhade) [inline content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md:30]:
```suggestion
```pseudo
```

File: content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+import torch.nn as nn
+import torch.optim as optim
+import torch.distributed as dist
+from torch.nn.parallel import DistributedDataParallel as DDP
+
+# Create a function to set up the initialisation process
+def setup(rank, world_size):
+    os.environ['MASTER_ADDR'] = 'localhost'
+    os.environ['MASTER_PORT'] = '8000'
+    dist.init_process("nccl", rank=rank, world_size=world_size)
```
Comment (Radhika-okhade) [inline content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md:74]:
```suggestion
    dist.init_process_group("nccl", rank=rank, world_size=world_size)
```

**Comment by Radhika-okhade** [comment]
Hey! @andersooi, Thank you for contributing to Codecademy docs. I have made a few suggestions; please go through them and make the necessary changes.

File: content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
+---
+Title: 'Distributed Data Parallelism'
+Description: 'An overview of distributed data parallelism in PyTorch on how it enables efficient training of large models across multiple GPUs.'
```
Comment (Radhika-okhade) [inline content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md:3]:
```suggestion
Description: 'Enables users to efficiently train models across multiple GPUs and machines.'
```

**Review by Radhika-okhade — APPROVED** [review]

File: content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+Tags:
+  - 'Data'
+  - 'Data Parallelism'
+  - 'PyTorch'
+CatalogContent:
+  - 'intro-to-py-torch-and-neural-networks'
+  - 'paths/build-a-machine-learning-model'
+---
+
+## Introduction to Distributed Data Parallelism
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md:16]: We can remove this heading as we usually start explaining the term as per our [style guide](https://github.com/Codecademy/docs/blob/main/documentation/style-guide.md).

```suggestion
```

File: content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+Distributed Data Parallelism (DDP) in PyTorch is a module that enables users to efficiently train models across multiple GPUs and machines. By splitting the training process across multiple machines, DDP helps reduce training time and facilitates scaling to larger models and datasets.
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md:18]: When adding a new term for the first time, we should emphasise it in bold.

```suggestion
**Distributed Data Parallelism (DDP)** in PyTorch is a module that enables users to train models across multiple GPUs and machines efficiently. By splitting the training process across multiple machines, DDP helps reduce training time and facilitates scaling to larger models and datasets.
```

File: content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+It achieves parallelism by splitting the input data into smaller chunks, processing them on different GPUs, and aggregating results for updates. Compared to `DataParallel`, DDP offers better performance and scalability by minimizing device communication overhead.
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md:19]: We should replace "splitting" with "partitioning" for technical precision.

```suggestion
DDP achieves parallelism by partitioning input data into smaller chunks, processing them on multiple GPUs, and aggregating the results for updates. Compared to `DataParallel`, DDP offers better performance and scalability by minimizing device communication overhead.
```

File: content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+### Key Features
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md:21]: This should be h2

```suggestion
## Key Features
```

File: content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+- High performance: Efficient communication using NVIDIA Collective Communications Library (NCCL) or Gloo backend (for Windows platforms).
+- Fault tolerance: Handles errors during distributed training.
+- Scalability: Suitable for multi-node, multi-GPU setups.
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md:25]: The list headers should be bold so that the reader quickly understands what is being explained.

```suggestion
- **High performance**: Enables efficient communication using _NVIDIA Collective Communications Library (NCCL)_ or _Gloo backend_ (for Windows platforms).
- **Fault tolerance**: Provides mechanisms to handle errors during distributed training.
- **Scalability**: Designed for seamless scaling in multi-node and multi-GPU environments.
```

File: content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+# Initialise the process group
+dist.init_process_group(backend="nccl", init_method="env://", world_size=4, rank=0)
+
+# Wrap the model with DDP
+model = torch.nn.Linear(10, 10).cuda()
+model = DDP(model)
+```
+
+- `backend`: This is optional and by default, a `gloo` and `nccl` backend will be created.
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md:49]: This can be better written like this -

```suggestion
- `backend` (Optional): Specifies the backend for distributed communication.
	- `nccl` (NVIDIA Collective Communications Library):
		- Optimized for NVIDIA GPUs.
		- Supports high-performance GPU communication using CUDA.
		- Preferred choice for GPU-based distributed training.
	- `gloo`:
		- A cross-platform backend.
		- Used for CPU-based training or on systems without GPUs.
```

File: content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+- `init_method`: This is optional as well and the default is `env://`.
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md:50]: This needs a better explanation of what it does and the different options available for this parameter.

```suggestion
- `init_method` (Optional): Specifies the method for initializing the process group.
	- `env://`:
		- Uses environment variables to initialize the process group.
		- Simple and commonly used.
	- `file://<path-to-file>`:
		- Uses a shared file to store initialization information.
		- Useful for setups where environment variables aren't feasible.
	- `tcp://<ip>:<port>`:
		- Uses a TCP address for initialization.
		- Typically used in manual multi-node setups.
```

File: content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+
+### Required Environment Variables
+
+- `WORLD_SIZE`: The total number of processes.
+- `RANK`: Rank of the current process and must be between 0 and WORLD_SIZE - 1.
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md:55]: The constants should be wrapped in backticks.

```suggestion
- `RANK`: Specifies the rank (unique ID) of the current process within the group and must be between `0` and `WORLD_SIZE - 1`.
```

File: content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+model = torch.nn.Linear(10, 10).cuda()
+model = DDP(model)
+```
+
+- `backend`: This is optional and by default, a `gloo` and `nccl` backend will be created.
+- `init_method`: This is optional as well and the default is `env://`.
+
+### Required Environment Variables
+
+- `WORLD_SIZE`: The total number of processes.
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md:54]: Explanation can be better like this -

```suggestion
- `WORLD_SIZE`: Defines the total number of processes participating in the distributed training.
```

File: content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+    # Dummy data and labels
+    data = torch.randn(20, 10).cuda()
+    labels = torch.randn(20, 10).cuda()
+
+    # Training
+    for _ in range(10):
+        optimiser.zero_grad()
+        outputs = ddp_model(data)
+        loss = nn.MSELoss()(outputs, labels)
+        loss.backward()
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md:98]: Add a print statement to get a better idea of what's happening in the example on running -

```suggestion
        loss = nn.MSELoss()(outputs, labels)
        print(f"Epoch {epoch + 1}, Loss: {loss.item():.4f}")
        loss.backward()
```

The add output for this code - 

When this code is executed, the output might look like this (exact values will vary due to randomness):

```shell
Epoch 1, Loss: 1.2345
Epoch 2, Loss: 0.8765
Epoch 3, Loss: 0.5432
```

This output demonstrates that the DDP setup functions correctly, distributing the data and computing gradients across GPUs. If the program were run with multiple processes (e.g., `world_size=4`), each process would handle its share of the workload.

**Review by PragatiVerma18 — CHANGES_REQUESTED** [review]
Hey @andersooi, thanks for contributing to docs. The entry looks good. I have a few suggestions. Please check them and make the required changes, and then we will be good to merge this PR. Let me know if you need any help.

**Review by PragatiVerma18 — APPROVED** [review]
Hey @andersooi, this entry is now good to be merged! 🚀
